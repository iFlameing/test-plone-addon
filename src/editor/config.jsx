import React from 'react';
import MarkButton from './components/MarkButton';
import BlockButton from './components/BlockButton';
import Separator from './components/Separator';

import boldIcon from '@plone/volto/icons/bold.svg';
import codeIcon from '@plone/volto/icons/code.svg';
import headingIcon from '@plone/volto/icons/heading.svg';
import italicIcon from '@plone/volto/icons/italic.svg';
import listBulletIcon from '@plone/volto/icons/list-bullet.svg';
import listNumberedIcon from '@plone/volto/icons/list-numbered.svg';
import subheadingIcon from '@plone/volto/icons/subheading.svg';
import underlineIcon from '@plone/volto/icons/underline.svg';

// TODO: correct the file name in Volto:
import strikethroughIcon from '@plone/volto/icons/strickthrough.svg';

export const availableButtons = {
  bold: (props) => <MarkButton format="bold" icon={boldIcon} {...props} />,
  italic: (props) => (
    <MarkButton format="italic" icon={italicIcon} {...props} />
  ),
  underline: (props) => (
    <MarkButton format="underline" icon={underlineIcon} {...props} />
  ),
  strikethrough: (props) => (
    <MarkButton format="strikethrough" icon={strikethroughIcon} {...props} />
  ),
  code: (props) => <MarkButton format="code" icon={codeIcon} {...props} />,
  'heading-two': (props) => (
    <BlockButton format="heading-two" icon={headingIcon} {...props} />
  ),
  'heading-three': (props) => (
    <BlockButton format="heading-three" icon={subheadingIcon} {...props} />
  ),
  'numbered-list': (props) => (
    <BlockButton format="numbered-list" icon={listNumberedIcon} {...props} />
  ),
  'bulleted-list': (props) => (
    <BlockButton format="bulleted-list" icon={listBulletIcon} />
  ),
  separator: (props) => <Separator />,
};

export const defaultToolbarButtons = [
  'bold',
  'italic',
  'underline',
  'strikethrough',
  'separator',
  'heading-two',
  'heading-three',
  'separator',
  'numbered-list',
  'bulleted-list',
];

export let toolbarButtons = [...defaultToolbarButtons];

export let expandedToolbarButtons = [...defaultToolbarButtons];

export const decorators = [];

export const hotkeys = {
  'mod+b': 'bold',
  'mod+i': 'italic',
  'mod+u': 'underline',
  'mod+`': 'code',
  // TODO: more hotkeys, including from plugins!
};

export const listTypes = ['numbered-list', 'bulleted-list'];

export const availableLeafs = {
  bold: ({ children }) => {
    return <strong>{children}</strong>;
  },
  code: ({ children }) => {
    return <code>{children}</code>;
  },
  italic: ({ children }) => <em>{children}</em>,
  underline: ({ children }) => <u>{children}</u>,
  strikethrough: ({ children }) => <s>{children}</s>,
};

export const leafs = ['bold', 'code', 'italic', 'underline', 'strikethrough'];

export const keyDownHandlers = {};
