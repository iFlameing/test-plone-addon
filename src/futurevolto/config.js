import ObjectWidget from './ObjectWidget';
import ObjectBrowserWidget from '@plone/volto/components/manage/Widgets/ObjectBrowserWidget';

const install = (config) => {
  config.widgets.widget.object = ObjectWidget;
  config.widgets.widget.object_browser = ObjectBrowserWidget;
  return config;
};

export default install;
