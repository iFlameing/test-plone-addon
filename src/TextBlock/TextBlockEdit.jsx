/**
 * A lot of inspiration from the great https://github.com/udecode/slate-plugins/,
 * especially the list element.
 */

import React, { useMemo, useState } from 'react';
import { connect } from 'react-redux';
import { Button } from 'semantic-ui-react';

import { Icon, BlockChooser, SidebarPortal } from '@plone/volto/components';
import addSVG from '@plone/volto/icons/circle-plus.svg';
import { settings } from '~/config';

import { setSlateBlockSelection } from './../actions';
import SlateEditor from './../editor';
import { plaintext_serialize } from './../editor/render';
import ShortcutListing from './ShortcutListing';
import { withList, withDeserializeHtml } from './decorators';
import {
  getBackspaceKeyDownHandlers,
  getFocusRelatedKeyDownHandlers,
  onKeyDownList,
} from './keyDownHandlers';

const TextBlockEdit = (props) => {
  const {
    block,
    blockNode,
    data,
    detached,
    index,
    onAddBlock,
    onChangeBlock,
    onDeleteBlock,
    onFocusNextBlock,
    onFocusPreviousBlock,
    onMutateBlock,
    onSelectBlock,
    properties,
    selected,
    setSlateBlockSelection,
  } = props;

  const { value } = data;
  const [addNewBlockOpened, setAddNewBlockOpened] = useState();

  const keyDownHandlers = useMemo(() => {
    return {
      ...getBackspaceKeyDownHandlers({
        block,
        onDeleteBlock,
        index,
        properties,
        setSlateBlockSelection,
        onChangeBlock,
        onFocusPreviousBlock,
        blockNode,
      }),
      ...getFocusRelatedKeyDownHandlers({
        block,
        blockNode,
        onFocusNextBlock,
        onFocusPreviousBlock,
      }),
      ...settings.slate?.keyDownHandlers,
    };
  }, [
    block,
    blockNode,
    index,
    onFocusNextBlock,
    onFocusPreviousBlock,
    onDeleteBlock,
    onChangeBlock,
    properties,
    setSlateBlockSelection,
  ]);

  const configuredWithList = useMemo(
    () => withList({ onChangeBlock, onAddBlock, onSelectBlock, index }),
    [index, onAddBlock, onChangeBlock, onSelectBlock],
  );

  const configuredOnKeyDownList = useMemo(() => onKeyDownList(), []);

  let timeoutTillRerender = null;
  React.useEffect(() => {
    return () => {
      if (timeoutTillRerender) {
        clearTimeout(timeoutTillRerender);
      }
    };
  });

  return (
    <>
      <SidebarPortal selected={selected}>
        <div id="slate-plugin-sidebar"></div>
        <ShortcutListing />
      </SidebarPortal>
      <SlateEditor
        index={index}
        properties={properties}
        onAddBlock={onAddBlock}
        decorators={[configuredWithList, withDeserializeHtml]}
        onSelectBlock={onSelectBlock}
        value={value}
        data={data}
        block={block}
        onChange={(value, selection) => {
          // without using setTimeout, the user types characters on the right side of the text cursor
          timeoutTillRerender = setTimeout(() => {
            setSlateBlockSelection(block, selection);
          });

          onChangeBlock(block, {
            ...data,
            value,
            plaintext: plaintext_serialize(value || []),
          });
        }}
        onKeyDown={({ editor, event }) => {
          configuredOnKeyDownList(event, editor);

          keyDownHandlers[event.key] &&
            keyDownHandlers[event.key]({
              ...props,
              editor,
              event,
            });
        }}
        selected={selected}
        placeholder="Enter some rich textâ€¦"
      />
      {!detached && !data.plaintext && (
        <Button
          basic
          icon
          onClick={() => setAddNewBlockOpened(!addNewBlockOpened)}
          className="block-add-button"
        >
          <Icon name={addSVG} className="block-add-button" size="24px" />
        </Button>
      )}
      {addNewBlockOpened && (
        <BlockChooser onMutateBlock={onMutateBlock} currentBlock={block} />
      )}
    </>
  );
};

export default connect(null, {
  setSlateBlockSelection,
})(TextBlockEdit);
