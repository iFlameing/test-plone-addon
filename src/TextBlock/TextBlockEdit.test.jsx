import React from 'react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-intl-redux';
import { render } from '@testing-library/react';

import TextBlockEdit from './TextBlockEdit';

// TODO: use this instead of expecting some props to be undefined
import { settings } from '~/config';

const mockStore = configureStore();

// ReactDOM.createPortal = (node) => node;

window.getSelection = () => null;

describe('TextBlockEdit', () => {
  it('renders w/o errors', async () => {
    const store = mockStore({
      intl: {
        locale: 'en',
        messages: {},
      },
    });

    // TODO: also test for the initial contents: My first paragraph.
    const { asFragment } = render(
      <Provider store={store}>
        <TextBlockEdit
          block="478923"
          blockNode={{ current: {} }}
          detached={false}
          index={2}
          onAddBlock={() => {}}
          onChangeBlock={() => {}}
          onDeleteBlock={() => {}}
          onFocusNextBlock={() => {}}
          onFocusPreviousBlock={() => {}}
          onMutateBlock={() => {}}
          onSelectBlock={() => {}}
          properties={{}}
          setSlateBlockSelection={() => {}}
          data={{
            value: [
              {
                type: 'paragraph',
                children: [{ text: '' /* My first paragraph. */ }],
              },
            ],
          }}
          selected={true}
        />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
