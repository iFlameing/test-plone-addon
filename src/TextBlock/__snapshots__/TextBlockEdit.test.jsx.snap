// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextBlockEdit renders w/o errors 1`] = `
<DocumentFragment>
  <div
    class="slate-editor selected"
  >
    <div
      class="toolbar-wrapper"
    />
    <div
      contenteditable="true"
      data-gramm="false"
      data-slate-editor="true"
      data-slate-node="value"
      role="textbox"
      style="outline: none; white-space: pre-wrap; word-wrap: break-word;"
    >
      <p
        data-slate-node="element"
      >
        <span
          data-slate-node="text"
        >
          <span
            data-slate-leaf="true"
          />
        </span>
      </p>
    </div>
  </div>
  <button
    class="ui basic icon button block-add-button"
  >
    <svg
      class="icon block-add-button"
      style="height: 24px; width: auto; fill: currentColor;"
      viewBox=""
      xmlns=""
    />
  </button>
</DocumentFragment>
`;
